Q1. Predict “ocean proximity” in the housing.csv file using both MLP Classifier and Keras. Use 50% of the data as the test set.

- Using the MLP Classifier, build and train a deep neural network with three hidden layers of 15, 10 and 7 nodes with max iterations set to 500. Do this with and without early stopping. In each case, compute and print the confusion matrix on the test set, as well as print the iteration at which the training terminated. Did early stopping help, or hurt?
- Use Keras to build a network of the same structure (i.e. nodes and hidden layers). In Keras it is OK if the other hyper-parameters are different and in Keras you can build only one model (with OR without early stopping – your choice). Print the confusion matrix on the test set (with or without early stopping – up to you). Was the model built in keras better or worse than what you got previously with MLP classifiers?
ONLY KEEP THE CODE THAT IS NEEDED

In [1]:
# Importing Libraries
import numpy as np
import os
import pandas as pd
import tensorflow as tf

In [2]:
#Loading the data
housing = pd.read_csv("housing.csv")
housing.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [3]:
#Dropping the missing values
housing.dropna(axis=0, inplace=True)

In [4]:
# Target Variables and Input variables
housing_X = housing.drop("ocean_proximity", axis=1) # drop labels 
housing_target = housing["ocean_proximity"]
housing_target.value_counts()

<1H OCEAN     9034
INLAND        6496
NEAR OCEAN    2628
NEAR BAY      2270
ISLAND           5
Name: ocean_proximity, dtype: int64

In [5]:
#Standardizing the data
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
housing_X_std = scaler.fit_transform(housing_X)

In [6]:
# Spliiting Train,Test
from sklearn.model_selection import train_test_split
train_x, test_x, train_y, test_y = train_test_split(housing_X_std, housing_target, test_size=0.5)

**MLP Classfier without early stopping**

In [7]:
from sklearn.neural_network import MLPClassifier
mlp_clf = MLPClassifier(hidden_layer_sizes=(15,10,7),max_iter=500,verbose=True)
mlp_clf.fit(train_x, train_y)

Iteration 1, loss = 1.75648782
Iteration 2, loss = 1.42839494
Iteration 3, loss = 1.13294112
Iteration 4, loss = 0.93219725
Iteration 5, loss = 0.80456934
Iteration 6, loss = 0.72593659
Iteration 7, loss = 0.67235852
Iteration 8, loss = 0.63295429
Iteration 9, loss = 0.59711400
Iteration 10, loss = 0.56438855
Iteration 11, loss = 0.53576706
Iteration 12, loss = 0.50892063
Iteration 13, loss = 0.48476988
Iteration 14, loss = 0.46070984
Iteration 15, loss = 0.44141533
Iteration 16, loss = 0.42452381
Iteration 17, loss = 0.40930021
Iteration 18, loss = 0.39549395
Iteration 19, loss = 0.38637101
Iteration 20, loss = 0.37562992
Iteration 21, loss = 0.36893765
Iteration 22, loss = 0.36205516
Iteration 23, loss = 0.35609232
Iteration 24, loss = 0.35016459
Iteration 25, loss = 0.34630005
Iteration 26, loss = 0.34141313
Iteration 27, loss = 0.33484797
Iteration 28, loss = 0.33113011
Iteration 29, loss = 0.32887186
Iteration 30, loss = 0.32362244
Iteration 31, loss = 0.31984259
Iteration 32, los

MLPClassifier(hidden_layer_sizes=(15, 10, 7), max_iter=500, verbose=True)

In [8]:
mlp_clf.n_iter_

288

We can see that the iterating has stopped at 288 as the training loss did not improve after the iteration at tolerance=0.000100 for 10 epochs.

In [9]:
test_y_pred = mlp_clf.predict(test_x)

In [10]:
from sklearn.metrics import confusion_matrix
confusion_matrix(test_y, test_y_pred)

array([[4025,   87,    0,   80,  249],
       [ 120, 3126,    0,   31,    0],
       [   0,    0,    0,    0,    2],
       [  41,    9,    0, 1057,   20],
       [ 258,    0,    0,   74, 1038]])

In [11]:
from sklearn.metrics import accuracy_score
print(accuracy_score(test_y, test_y_pred))

0.9049623177057845


The accuarcy for the MLP classifier without early stopping is 90.4%.

**MLP with Early Stopping**

In [12]:
mlp_clf2 = MLPClassifier(hidden_layer_sizes=(15,10,7),max_iter=500,verbose=True,early_stopping=True,)
mlp_clf2.fit(train_x, train_y)

Iteration 1, loss = 1.97229319
Validation score: 0.273973
Iteration 2, loss = 1.63356611
Validation score: 0.394325
Iteration 3, loss = 1.34687098
Validation score: 0.635029
Iteration 4, loss = 1.16186527
Validation score: 0.670254
Iteration 5, loss = 1.03570155
Validation score: 0.699609
Iteration 6, loss = 0.93238227
Validation score: 0.722114
Iteration 7, loss = 0.84193039
Validation score: 0.738748
Iteration 8, loss = 0.76347143
Validation score: 0.761252
Iteration 9, loss = 0.69625271
Validation score: 0.768102
Iteration 10, loss = 0.64221531
Validation score: 0.774951
Iteration 11, loss = 0.59361734
Validation score: 0.777886
Iteration 12, loss = 0.54843718
Validation score: 0.792564
Iteration 13, loss = 0.51156096
Validation score: 0.805284
Iteration 14, loss = 0.48120209
Validation score: 0.817025
Iteration 15, loss = 0.45498912
Validation score: 0.819961
Iteration 16, loss = 0.43413019
Validation score: 0.827789
Iteration 17, loss = 0.41763082
Validation score: 0.832681
Iterat

MLPClassifier(early_stopping=True, hidden_layer_sizes=(15, 10, 7), max_iter=500,
              verbose=True)

In [13]:
mlp_clf2.n_iter_

71

In [14]:
test_y_pred = mlp_clf2.predict(test_x)
from sklearn.metrics import confusion_matrix
confusion_matrix(test_y, test_y_pred)

array([[4004,   45,    0,  137,  255],
       [ 185, 3060,    0,   31,    1],
       [   0,    0,    0,    0,    2],
       [ 149,   28,    0,  902,   48],
       [ 301,    0,    0,   37, 1032]])

In [15]:
from sklearn.metrics import accuracy_score
print(accuracy_score(test_y, test_y_pred))

0.8806890476656553


We can see that the iterating has stopped at 71 and the accuracy is 88.0%

From the above, we can say that the early stopping did not help here as the accuracy did not increase.

**KERAS MODEL**

In [75]:
housing2 = pd.read_csv("housing.csv")
#Drop the missing values
housing2.dropna(axis=0, inplace=True)

In [76]:
#Drop the missing values
housing2.dropna(axis=0, inplace=True)
housing2.reset_index(inplace=True, drop=True)
housing2.describe()

longitude      latitude  housing_median_age   total_rooms  \
count  20433.000000  20433.000000        20433.000000  20433.000000   
mean    -119.570689     35.633221           28.633094   2636.504233   
std        2.003578      2.136348           12.591805   2185.269567   
min     -124.350000     32.540000            1.000000      2.000000   
25%     -121.800000     33.930000           18.000000   1450.000000   
50%     -118.490000     34.260000           29.000000   2127.000000   
75%     -118.010000     37.720000           37.000000   3143.000000   
max     -114.310000     41.950000           52.000000  39320.000000   

       total_bedrooms    population    households  median_income  \
count    20433.000000  20433.000000  20433.000000   20433.000000   
mean       537.870553   1424.946949    499.433465       3.871162   
std        421.385070   1133.208490    382.299226       1.899291   
min          1.000000      3.000000      1.000000       0.499900   
25%        296.000000    787.000000    280.000000       2.563700   
50%        435.000000   1166.000000    409.000000       3.536500   
75%        647.000000   1722.000000    604.000000       4.744000   
max       6445.000000  35682.000000   6082.000000      15.000100   

       median_house_value  
count        20433.000000  
mean        206864.413155  
std         115435.667099  
min          14999.000000  
25%         119500.000000  
50%         179700.000000  
75%         264700.000000  
max         500001.000000

In [77]:
housing_num = housing2.drop("ocean_proximity", axis=1)
housing_target = housing2[["ocean_proximity"]]

In [78]:
from sklearn.preprocessing import StandardScaler
scaler = StandardScaler()
housing_num_std = scaler.fit_transform(housing_num)

In [64]:
from sklearn.preprocessing import OrdinalEncoder
ordinal_encoder = OrdinalEncoder()
housing_labels_ord = ordinal_encoder.fit_transform(housing_target)
housing_labels_ord[:10]

array([[3.],
       [3.],
       [3.],
       [3.],
       [3.],
       [3.],
       [3.],
       [3.],
       [3.],
       [3.]])

In [65]:
housing_labels_int = housing_labels_ord.astype(int)
housing_labels_int.dtype

dtype('int64')

In [79]:
from sklearn.model_selection import train_test_split
train_x, test_x, train_y, test_y = train_test_split(housing_num_std, housing_labels_int, test_size=0.5)

In [80]:
from tensorflow import keras
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.layers import Dropout

In [81]:
model = Sequential()
model.add(Dense(15, input_dim=9, activation='relu'))
model.add(Dense(10, activation='relu'))
model.add(Dense(7, activation='relu'))
model.add(Dense(5, activation='softmax'))
sgd = keras.optimizers.SGD(lr=0.05)
model.compile(loss='sparse_categorical_crossentropy', optimizer=sgd, metrics=['accuracy'])
model.fit(train_x, train_y, epochs=100, batch_size=100)

Epoch 1/100
103/103 [==============================] - 0s 730us/step - loss: 1.2337 - accuracy: 0.5571
Epoch 2/100
103/103 [==============================] - 0s 608us/step - loss: 0.9880 - accuracy: 0.6320
Epoch 3/100
103/103 [==============================] - 0s 577us/step - loss: 0.8290 - accuracy: 0.6833
Epoch 4/100
103/103 [==============================] - 0s 536us/step - loss: 0.7180 - accuracy: 0.7219
Epoch 5/100
103/103 [==============================] - 0s 565us/step - loss: 0.6377 - accuracy: 0.7616
Epoch 6/100
103/103 [==============================] - 0s 539us/step - loss: 0.5785 - accuracy: 0.7837
Epoch 7/100
103/103 [==============================] - 0s 662us/step - loss: 0.5240 - accuracy: 0.7898
Epoch 8/100
103/103 [==============================] - 0s 653us/step - loss: 0.4567 - accuracy: 0.8093
Epoch 9/100
103/103 [==============================] - 0s 583us/step - loss: 0.4147 - accuracy: 0.8356
Epoch 10/100
103/103 [==============================] - 0s 561us/step - l

In [82]:
# evaluate the model
scores = model.evaluate(test_x, test_y)
print("\n%s: %.2f%%" % (model.metrics_names[1], scores[1]*100))

320/320 [==============================] - 0s 519us/step - loss: 0.2296 - accuracy: 0.9026

accuracy: 90.26%


We can say that the MLP model and keras model performed almost same as the accuracy is around 90.4 % and the accuracy for the keras model is 90.2%.